### PR TITLE
SOLR-15315: contrib/ltr MultipleAdditiveTreesModel latency optimisation by changing recursion to loop

### PR DESCRIPTION
**Pull request by suryakant261**
# Description

LTR tree traversal currently happens in a recursive way.

Background :When we tried to increase the number of trees in our production solr, we were facing latency issue. From solr profiling it came out that most time is spent in recursive call for tree traversal.

# Solution

Recursion is slower due to overhead of function call stack, so we optimised this and translated it to loop based approach. This improved solr latency significantly and we were able to scale much better.

# Tests

Have run internal load tests and compared against master branch.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes. (Not required, non-functional optimisation)
- [ ] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

### CODE REVIEW

**Comment by suryakant261** [comment]
@madrob @janhoy  Please review

**Comment by suryakant261** [comment]
@dsmiley @dweiss @bruno-roustant  Please review

**Comment by dsmiley** [comment]
You are tagging people that have not touched this code :-). I tagged @cpoerschke (who does touch it) for review.

**Comment by suryakant261** [comment]
> You are tagging people that have not touched this code :-). I tagged @cpoerschke (who does touch it) for review.

Thank you :)

**Comment by suryakant261** [comment]
@cpoerschke @ErickErickson @romseygeek @michaelbraun @dweiss Please review

File: solr/contrib/ltr/src/java/org/apache/solr/ltr/model/MultipleAdditiveTreesModel.java
Comment (alessandrobenedetti) [inline solr/contrib/ltr/src/java/org/apache/solr/ltr/model/MultipleAdditiveTreesModel.java:173]: RegressionTreeNode regressionTreeNode 
I am usually not a big fan of type = variable name approach, not much information is added with the name
maybe a simple "current" ?

**Review by alessandrobenedetti — APPROVED** [review]
removing recursive calls kinda defeats the meaning of having the RegressionNode to have the score method, potentially only the root needs that, anyway it just a consideration,
 I am ok with this, if all tests are green and we get performance improvements with such a quick one, let' go for it!
Waiting for @cpoerschke approval and we can merge.
it's also needed to add the Jira issue as a bugfix in the changes (I don't like this, because it's easy to forget, we should probably find someway for that to be automatic (unless it has been changed)

**Comment by cpoerschke** [comment]
Sorry, I'd put this on my TODO list for this week but didn't say so here, only on the JIRA ticket.

Thanks @alessandrobenedetti for also taking a look!

I agree it makes sense to remove the recursion in the `score` method if that improves latency. Any thoughts on also removing recursion in the `explain` method, since they are related and then would follow the same approach?

**Comment by cpoerschke** [comment]
>  removing recursive calls kinda defeats the meaning of having the RegressionNode to have the score method, potentially only the root needs that, anyway it just a consideration, ...

And not having the method would also remove the potential for implementation mishaps such as `threshold` i.e. `this.threshold` being used instead of `regressionTreeNode.threshold` in the loop, or for code readers to need to consider that those three possibilities exist. Added a commit to try this out, feel free to revert or adapt. What do you think?

**Comment by suryakant261** [comment]
Hi @alessandrobenedetti  and @cpoerschke 

Really thanks for reviewing this.  Have added an entry in CHANGES.txt and made the same recursion to loop changes for explain and validate methods as well.

File: solr/CHANGES.txt
Comment (alessandrobenedetti) [inline solr/CHANGES.txt:131]: Add Christine in here as well, please :) she helped with some code and review effort, it's good to give merit

**Review by alessandrobenedetti — APPROVED** [review]
It looks much better to me, just a tiny note on the CHANGES.txtm but for me it's good to go

File: solr/CHANGES.txt
Comment (cpoerschke) [reply solr/CHANGES.txt]: Alessandro too I think :)

Added https://github.com/apache/solr/pull/58/commits/d522b04186783a0d451d5c3f70e2fc9cef0f4e33 edit (details in the commit message) and will merge shortly.

**Comment by cpoerschke** [comment]
> Makes sense to keep like this to avoid errors coming because of this. vs regressionTreeNode. Followed the same approach in validate and explain methods

Good catch about `validate` also having recursion, thanks @suryakant261 for including that in the PR iteration!

File: solr/contrib/ltr/src/java/org/apache/solr/ltr/model/MultipleAdditiveTreesModel.java
Comment (cpoerschke) [inline solr/contrib/ltr/src/java/org/apache/solr/ltr/model/MultipleAdditiveTreesModel.java:322]:
```suggestion
      if (null == regressionTreeNode.left) {
        throw new ModelException("MultipleAdditiveTreesModel tree node is missing left");
      } else {
        regressionTreeNode = regressionTreeNode.left;
        continue;
      }
      if (null == regressionTreeNode.right) {
        throw new ModelException("MultipleAdditiveTreesModel tree node is missing right");
      } else {
        regressionTreeNode = regressionTreeNode.right;
        continue;
      }
```

I think a `continue` is needed after the `regressionTreeNode = regressionTreeNode.left` and whilst it's not necessary currently after the `regressionTreeNode = regressionTreeNode.right` it might be nicely symmetrical to have it there too e.g. in case additional validation is added in future after the `regressionTreeNode = regressionTreeNode.right` assignment.

File: solr/contrib/ltr/src/java/org/apache/solr/ltr/model/MultipleAdditiveTreesModel.java
Comment (cpoerschke) [inline solr/contrib/ltr/src/java/org/apache/solr/ltr/model/MultipleAdditiveTreesModel.java:328]: minor:
```suggestion
    final StringBuilder returnValueBuilder = new StringBuilder();
    while (true) {
```

**Comment by cpoerschke** [comment]
Hmm, running locally the tests don't pass for me via `./gradlew -p solr/contrib/ltr test` command. The pull request CI does not automatically run these tests.

**Review by cpoerschke — CHANGES_REQUESTED** [review]
Hi @suryakant261, could I ask you to do the following please:
* revisit the `validate` change to ensure all of the tree continues to be validated
* run the `solr/contrib/ltr` tests locally to check they pass

When you get a chance, no rush, we all contribute here as and when time permits. Thanks!

File: solr/contrib/ltr/src/java/org/apache/solr/ltr/model/MultipleAdditiveTreesModel.java
Comment (suryakant261) [reply solr/contrib/ltr/src/java/org/apache/solr/ltr/model/MultipleAdditiveTreesModel.java]: Good catch @cpoerschke , thanks for highlighting. In Explain and score method, we want to  traverse only one path from root to leaf based on featurevalue, however in validate method we want to traverse all nodes of the tree. So, the approach had to be different,  have now  updated the approach for validate method, using a stack and doing preorder traversal.

**Comment by suryakant261** [comment]
> Hi @suryakant261, could I ask you to do the following please:
> 
> * revisit the `validate` change to ensure all of the tree continues to be validated
> * run the `solr/contrib/ltr` tests locally to check they pass
> 
> When you get a chance, no rush, we all contribute here as and when time permits. Thanks!

1. Done using iterative preorder traversal.
2. All the test cases passing in the local run now. (Wasn't aware that these are not part of Pull requests build, should we enable it?)
image

File: solr/contrib/ltr/src/java/org/apache/solr/ltr/model/MultipleAdditiveTreesModel.java
Comment (cpoerschke) [inline solr/contrib/ltr/src/java/org/apache/solr/ltr/model/MultipleAdditiveTreesModel.java:317]:
```suggestion
        continue;
      }
```

or

```suggestion
      } else {
        ...
      }
```

Comment (suryakant261) [reply solr/contrib/ltr/src/java/org/apache/solr/ltr/model/MultipleAdditiveTreesModel.java]: done

**Comment by suryakant261** [comment]
Hi @cpoerschke ,
Could you please take a look, all the suggested changes are done with unit tests passing locally.

**Review by cpoerschke — APPROVED** [review]

**Comment by cpoerschke** [comment]
> Gradle Precommit / gradle precommit w/ Java 11 (pull_request)

This failed on the pull request here but succeeded locally, hmm. I will optimistically proceed to merge here and then watch out for any issues after merge before cherry-picking from `solr/main` to `lucene-solr/branch_8x` branch likely tomorrow.

**Pull request merged**
